core/src/main/resources/scala-native/socket.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2013-2018 EPFL
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef __linux__
+#define _GNU_SOURCE
+#include <errno.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+
+typedef unsigned short scalanative_sa_family_t;
+struct scalanative_sockaddr {
```

Comment:
I was just dealing with this today.  You wonder where my
time goes ;-)

I am not quite catching why this "glue" is necessary, but
that is probably lack of cleverness on my end.

I believe that you want to allocate the largest size possible that
 accept4 can possibly fill in.  Generally, one would use
a sockaddr_storage, defined specifically for that purpose.
That is not usable on 0.4.n (I have not checked 0.4.7 to
see what the merge elf did).  So, I suggest sockaddr_in6
and the adjusted length.   

One is supposed to do a check after accept to see if the
addrlen passed back exceeded the addrlen passed in.
In this context, sockaddr_in6 is the largest that can be
passed back, so that check can be skipped. However,
it is good to know about that behavior. (if the return size
is larger, that means the buffer passed in was too small
and got truncated/mangled).

Reply:
This is one of the mysteries of SN 0.4.x. I believe it is necessary due to the use of an intermediate `scalanative_sockaddr`?

For what it is worth, this code is entirely cribbed from Scala Native itself except switched to use `accept4` instead of `accept`. So any improvements should ideally be done in tandem, as we rely on the Scala Native version of this code when calling `accept` on macOS.

https://github.com/scala-native/scala-native/blob/2ef67fe81c2a71de7e8a9f57229dbdcb5e178998/posixlib/src/main/resources/scala-native/sys/socket.c#L226-L255

Reply:
After years of sustained effort, I finally eliminated "glue" in straight passthru  code in SN 0.5.0.
There is some code there to check, at compile time, that the structures are compatible.
IIRC, accept() was one of the ones which were simplified..

Reply:
I believe this file can go away completely, with
the corresponding change in socket.scala.

If you want to be extremely conservative, 
the "@name" and the code here can go away, and this
file becomes just the _Static_assert statements relative
to `struct sockaddr`.    

The 0.5.0 code which does away with these conversions
has been merged into the SN mainline and seem to 
have caused no problems.

Reply:
Hm, are you sure? That's what I at first and it was not working actually. I had to add this glue code to fix it.

Reply:
"Not working"? in what way?  

For the sake of expediency, I suggest the merge proceed
with the file and I will do private builds afterwards, from a
stable baseline, to see why reality is not matching my
expectations.  It has a habit of doing that, and pulling
my leash hard.

If there is something wrong with removing the file, there may be
an issue in 0.5.0.

Sorry for taking time.

Reply:
Not at all, I would like to get to the bottom of this and would be happy to see the glue code go away. See https://github.com/armanbilge/epollcat/pull/53.